### PR TITLE
Action: add release workflow

### LINKED ISSUE #11: Release on github action
- Semantic version.
- Build release image.

### PR DESCRIPTION
**Pull request by imeoer**
Add github action configuration to release acceld, accelctl binaries and image for the committed tag.

A successful release job can be found here: https://github.com/imeoer/acceleration-service/actions/runs/2164975305

Close #11

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+          tar cf - harbor-acceld | gzip > ${tarball}
+          echo "tag=$tag" >> $GITHUB_ENV
+          echo "tarball=$tarball" >> $GITHUB_ENV
+      - name: Release
+        uses: softprops/action-gh-release@v1
+        with:
+          name: "harbor-acceld ${{ env.tag }} release"
+          generate_release_notes: true
+          files:
+            ${{ env.tarball }}
```
Comment (ktock) [inline .github/workflows/release.yml:111]: Why not use `hub`? (`hub release create`)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          name: harbor-acceld-artifacts
+          path: harbor-acceld
+      - name: Upload Artifacts
+        run: |
+          tag=$(echo $GITHUB_REF | cut -d/ -f3-)
+          tarball="harbor-acceld-$tag-x86_64.tgz"
+          chmod +x harbor-acceld/*
+          tar cf - harbor-acceld | gzip > ${tarball}
+          echo "tag=$tag" >> $GITHUB_ENV
+          echo "tarball=$tarball" >> $GITHUB_ENV
```
Comment (ktock) [inline .github/workflows/release.yml:104]: Can we also upload the sha256sum of the release artifacts?

File: script/release/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+RUN make -C /accel
+
+# Install containerd
+RUN wget https://github.com/containerd/containerd/releases/download/v$CONTAINERD_VERSION/containerd-$CONTAINERD_VERSION-linux-amd64.tar.gz
+RUN mkdir -p /usr/local/bin/ && tar xzvf containerd-$CONTAINERD_VERSION-linux-amd64.tar.gz -C /usr/local/
+RUN ls -l /usr/local/bin/
+
+# Install nydus component
+RUN wget https://github.com/dragonflyoss/image-service/releases/download/$NYDUS_BUILER_VERSION/nydus-static-$NYDUS_BUILER_VERSION-x86_64.tgz
+RUN tar xzvf nydus-static-$NYDUS_BUILER_VERSION-x86_64.tgz && mv nydus-static/nydus-image /usr/local/bin/.
```
Comment (ktock) [inline script/release/Dockerfile:18]: Can we check the hash (sha256sum) of the tarballs?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go
+      - name: Build Binary
+        run: |
+          curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.43.0
+          export PATH=$PATH:$(go env GOPATH)/bin
+          make
```
[the remark this answers is not in the document]
Comment (imeoer) [reply .github/workflows/release.yml]: Thanks for the reply, fixed.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    runs-on: ubuntu-latest
+    needs: [release_binary]
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install Hub
+        run: |
+          HUB_VER=$(curl -s "https://api.github.com/repos/github/hub/releases/latest" | jq -r .tag_name | sed 's/^v//')
+          wget -q -O- https://github.com/github/hub/releases/download/v$HUB_VER/hub-linux-amd64-$HUB_VER.tgz | \
+          tar xz --strip-components=2 --wildcards '*/bin/hub'
+          sudo mv hub /usr/local/bin/hub
```
[the remark this answers is not in the document]
Comment (imeoer) [reply .github/workflows/release.yml]: Fixed

File: .github/workflows/release.yml (hunk 1 above)
Comment (imeoer) [reply .github/workflows/release.yml]: Fixed

File: .github/workflows/release.yml (hunk 2 above)
Comment (imeoer) [reply .github/workflows/release.yml]: Fixed

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        TAG=$GITHUB_REF_NAME
+        docker run -v $PWD/misc/config/config.yaml.nydus.tmpl:/etc/acceld-config.yaml -it -d --rm -p 2077:2077 goharbor/harbor-acceld:$TAG /etc/acceld-config.yaml
+        sleep 5
+        curl -f http://127.0.0.1:2077/api/v1/health
+
+    - name: Login Registry
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKER_HUB_USERNAME }}
+        password: ${{ secrets.DOCKER_HUB_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (imeoer) [reply .github/workflows/release.yml:40]: The harbor maintainer has already helped configure the secrets.

File: script/release/Dockerfile (hunk 3 above)
Comment (imeoer) [reply script/release/Dockerfile]: Thanks for the suggestion, will add the sha256sum file for nydus release first.

Comment (imeoer) [reply script/release/Dockerfile]: @ktock Fixed, added sha256sum check.

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      run: |
+        TAG=$GITHUB_REF_NAME
+        docker build -t goharbor/harbor-acceld:$TAG -f script/release/Dockerfile .
+
+    - name: Test Image
+      # Enable tty for docker
+      shell: 'script -q -e -c "bash {0}"'
+      run: |
+        TAG=$GITHUB_REF_NAME
+        docker run -v $PWD/misc/config/config.yaml.nydus.tmpl:/etc/acceld-config.yaml -it -d --rm -p 2077:2077 goharbor/harbor-acceld:$TAG /etc/acceld-config.yaml
```
Comment (ktock) [inline .github/workflows/release.yml:32]: Why `-it`?

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        username: ${{ secrets.DOCKER_HUB_USERNAME }}
+        password: ${{ secrets.DOCKER_HUB_PASSWORD }}
+
+    - name: Push Image
+      run: |
+        TAG=$GITHUB_REF_NAME
+        # Only push for non pull_request
+        if [ "$GITHUB_EVENT_NAME" != "pull_request" ]; then
+          docker push goharbor/harbor-acceld:$TAG
+        fi
```
Comment (ktock) [inline .github/workflows/release.yml:48]: How about using https://github.com/docker/build-push-action ?

File: script/release/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ARG NYDUS_BUILER_VERSION=v2.0.0-rc.6
+
+# Install acceld
+ADD ./ /accel
+RUN make -C /accel install-check-tools
+RUN make -C /accel
+
+# Install containerd
+RUN wget https://github.com/containerd/containerd/releases/download/v$CONTAINERD_VERSION/containerd-$CONTAINERD_VERSION-linux-amd64.tar.gz
+RUN wget https://github.com/containerd/containerd/releases/download/v$CONTAINERD_VERSION/containerd-$CONTAINERD_VERSION-linux-amd64.tar.gz.sha256sum
```
Comment (ktock) [inline script/release/Dockerfile:13]: checksum should be hardcoded to Dockerfile or some other files.

Comment (imeoer) [reply script/release/Dockerfile:13]: What is the purpose of sha256sum being hardcoded to Dockerfile, is there a reference example?

File: .github/workflows/release.yml (hunk 7 above)
Comment (imeoer) [reply .github/workflows/release.yml]: Fixed.

File: .github/workflows/release.yml (hunk 8 above)
Comment (imeoer) [reply .github/workflows/release.yml]: Updated, but my concern is that we need to do a simple test before pushing the image, and this action doesn't seem to do that?

File: script/release/Dockerfile (hunk 9 above)
Comment (ktock) [reply script/release/Dockerfile:13]: For checking that the downloaded binary is exactly same as the expected one in the Dockerfile.

kind example: https://github.com/kubernetes-sigs/kind/blob/main/images/base/Dockerfile#L30-L36
nerdctl example: https://github.com/containerd/nerdctl/tree/master/Dockerfile.d/SHA256SUMS.d

File: .github/workflows/release.yml (hunk 8 above)
Comment (imeoer) [reply .github/workflows/release.yml]: Thanks, fixed.

File: script/release/Dockerfile (hunk 9 above)
Comment (imeoer) [reply script/release/Dockerfile]: Thanks, fixed.

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "makefile.extensionOutputFolder": "./.vscode"
+}
```
[the remark this answers is not in the document]
Comment (imeoer) [reply .vscode/settings.json]: Fixed

File: script/release/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,33 @@
+FROM golang:1.17.7 AS build
+
+ARG CONTAINERD_VERSION=1.5.7
+ARG NYDUS_VERSION=v2.0.0-rc.6
+
+ARG CONTAINERD_LINUX_AMD64_SHA256SUM="109fc95b86382065ea668005c376360ddcd8c4ec413e7abe220ae9f461e0e173"
+ARG NYDUS_LINUX_AMD64_SHA256SUM="caca4be643ee11e74ff7be8409fc3bc0fff231395f324f41baadc3737a1156ec"
+
+# Install acceld
+ADD ./ /accel
```
Comment (ktock) [inline script/release/Dockerfile:10]: nit: `COPY` should be used when possible https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

File: script/release/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+RUN wget https://github.com/dragonflyoss/image-service/releases/download/$NYDUS_VERSION/nydus-static-$NYDUS_VERSION-linux-amd64.tgz
+RUN echo "$NYDUS_LINUX_AMD64_SHA256SUM  nydus-static-$NYDUS_VERSION-linux-amd64.tgz" | tee nydus-static-$NYDUS_VERSION-linux-amd64.tgz.sha256sum
+RUN sha256sum -c nydus-static-$NYDUS_VERSION-linux-amd64.tgz.sha256sum
+RUN tar xzvf nydus-static-$NYDUS_VERSION-linux-amd64.tgz && mv nydus-static/nydus-image /usr/local/bin/.
+
+# Build acceld image
+FROM photon:4.0
+COPY --from=build /accel/acceld /accel/accelctl /usr/local/bin/containerd /usr/local/bin/nydus-image /usr/local/bin/
+RUN mkdir /etc/containerd && /usr/local/bin/containerd config default > /etc/containerd/config.toml
+ADD ./script/release/entrypoint.sh /entrypoint.sh
```
Comment (ktock) [inline script/release/Dockerfile:31]: nit: `COPY` should be used when possible https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

**Review by ktock — COMMENTED** [review]
Overall looks good, leaving minor nits

File: script/release/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+RUN ls -l /usr/local/bin/
+
+# Install nydus component
+RUN wget https://github.com/dragonflyoss/image-service/releases/download/$NYDUS_VERSION/nydus-static-$NYDUS_VERSION-linux-amd64.tgz
+RUN echo "$NYDUS_LINUX_AMD64_SHA256SUM  nydus-static-$NYDUS_VERSION-linux-amd64.tgz" | tee nydus-static-$NYDUS_VERSION-linux-amd64.tgz.sha256sum
+RUN sha256sum -c nydus-static-$NYDUS_VERSION-linux-amd64.tgz.sha256sum
+RUN tar xzvf nydus-static-$NYDUS_VERSION-linux-amd64.tgz && mv nydus-static/nydus-image /usr/local/bin/.
+
+# Build acceld image
+FROM photon:4.0
```
Comment (ktock) [inline script/release/Dockerfile:28]: nit: Why not ubuntu?

Comment (ktock) [reply script/release/Dockerfile:28]: We're running tests on ubuntu so I think the image should be based on ubuntu too. https://github.com/goharbor/acceleration-service/blob/1c790244517ddb93f78d50313f140d2ebed9195e/.github/workflows/unit-test.yml#L11

File: script/release/Dockerfile (hunk 11 above)
Comment (imeoer) [reply script/release/Dockerfile]: Fixed, thanks!

File: script/release/Dockerfile (hunk 13 above)
Comment (imeoer) [reply script/release/Dockerfile:28]: https://github.com/goharbor/harbor/blob/main/make/photon/registry/Dockerfile.base

**Review by ktock — APPROVED** [review]

**Pull request merged**